src/common/com/intellij/plugins/haxe/model/type/HaxeTypeResolver.java

```diff
@@ -29,50 +30,62 @@
 import java.util.*;
 
 public class HaxeTypeResolver {
+  @NotNull
+  static public ResultHolder getFieldOrMethodReturnType(@NotNull AbstractHaxeNamedComponent comp) {
+    return getFieldOrMethodReturnType(comp, null);
+  }
+
   // @TODO: Check if cache works
-  static public SpecificTypeReference getFieldOrMethodReturnType(AbstractHaxeNamedComponent comp) {
+  @NotNull
+  static public ResultHolder getFieldOrMethodReturnType(@NotNull AbstractHaxeNamedComponent comp, @Nullable HaxeGenericResolver resolver) {
     // @TODO: cache should check if any related type has changed, which return depends
-    if (comp == null || comp.getContainingFile() == null) {
-      return SpecificHaxeClassReference.getUnknown(comp);
+    if (comp.getContainingFile() == null) {
+      return SpecificHaxeClassReference.getUnknown(comp).createHolder();
     }
     long stamp = comp.getContainingFile().getModificationStamp();
     if (comp._cachedType == null || comp._cachedTypeStamp != stamp) {
-      comp._cachedType = _getFieldOrMethodReturnType(comp);
+      comp._cachedType = _getFieldOrMethodReturnType(comp, resolver);
       comp._cachedTypeStamp = stamp;
     }
 
     return comp._cachedType;
   }
 
-  static public SpecificFunctionReference getMethodFunctionType(PsiElement comp) {
+  @NotNull
+  static public ResultHolder getMethodFunctionType(PsiElement comp, @Nullable HaxeGenericResolver resolver) {
     if (comp instanceof HaxeMethod) {
-      return ((HaxeMethod)comp).getModel().getFunctionType();
+      return ((HaxeMethod)comp).getModel().getFunctionType(resolver).createHolder();
     }
-    return null;
+    // @TODO: error
+    return SpecificTypeReference.getInvalid(comp).createHolder();
   }
 
-  static private SpecificTypeReference _getFieldOrMethodReturnType(AbstractHaxeNamedComponent comp) {
+  @NotNull
+  static private ResultHolder _getFieldOrMethodReturnType(AbstractHaxeNamedComponent comp, @Nullable HaxeGenericResolver resolver) {
```

Comment:
It is not used right now. But I think it will be required when improved it. At least with the implemented approach and I think I will need to evolve it a bit when required.
At this point, the implementation is a bit dummy. And it is not covering all the cases.
So that's why.